content/blog/symbolic-differentiation.Rmd

```diff
@@ -0,0 +1,396 @@
+
+```
+eval(differentiate(quote(-2 * x / (x * x - 3 * x)), "x"), list(x = pi))
+eval(D(quote(-2 * x / (x * x - 3 * x)), "x"), list(x = pi))
+```
+
+## Extending the implementation by adding more rules
+
+We could extend this easily now by adding more rules, and the implementation will even tell us what we need to add. So if we try and evaluate
+
+```{r, error = TRUE}
+differentiate(quote(exp(2 * x)), "x")
+```
+
+we get told to we need to implement the rule for `exp` which is simply:
+
+```{r}
+d_exp <- function(expr, name) {
+  call("*", differentiate(expr[[2]], name), expr)
+}
+```
+
+(that is, `d/dx exp(f(x))` is `f'(x) exp(f(x)))`. We add this to our set of rules:
+
+```{r}
+rules$exp <- d_exp
+```
+
+and now we can differentiate this new expression:
+
+```{r}
+differentiate(quote(exp(2 * x)), "x")
+```
+
+## Improving the implementation by writing sensible expressions
+
+Simplifying the expressions turns out to be much more work than the differentiation. The trick that we use is to avoid using `call()` directly to build expressions and create a simplifying expression builder that applies some simple rules to avoid building overly complicated expressions.  This does not simplify everything, but cuts out the most egregious bits of noise.
+
+This is probably not important for the efficiency of generated code (we're going to send this to an optimising compiler via some C++ code eventually) but it does make the resulting expressions easier to think about.
+
+Consider replacing `call("+", a, b)` with something that will avoid creating silly expressions.  If given numeric arguments for `a` and `b` we should sum them, and if either of `a` or `b` is zero we should return the other argument:
+
+```{r}
+m_plus <- function(a, b) {
+  if (is.numeric(a) && is.numeric(b)) {
+    a + b
+  } else if (is.numeric(b)) {
+    m_plus(b, a)
+  } else if (is.numeric(a) && a == 0) {
+    b
+  } else {
+    call("+", a, b)
+  }
+}
+```
+
+The pattern here is that only the last branch gives up and actually builds an expression with `call()`
+
+```{r}
+m_plus(3, 4)
+m_plus(quote(a), 0)
+m_plus(quote(a), 1)
+m_plus(0, quote(b))
+m_plus(1, quote(b))
+m_plus(quote(a), quote(b))
+```
+
+We can do similar things with multiplication:
+
+```{r}
+m_product <- function(a, b) {
+  if (is.numeric(a) && is.numeric(b)) {
+    a * b
+  } else if (is.numeric(b)) {
+    m_product(b, a)
+  } else if (is.numeric(a) && a == 0) {
+    0
+  } else if (is.numeric(a) && a == 1) {
+    b
+  } else {
+    call("*", a, b)
+  }
+}
+m_product(3, 4)
+m_product(quote(a), 0)
+m_product(quote(a), 1)
+m_product(quote(a), 2)
+m_product(quote(a), quote(b))
+```
+
+unary minus
+
+```{r}
+is_call <- function(x, name) {
+  is.recursive(x) && as.character(x[[1]]) == name
+}
+m_uminus <- function(a) {
+  if (is.numeric(a)) {
+    -a
+  } else if (length(x) == 2 && identical(x[[1]], quote(`-`))) {
+    a[[2]]
+  } else if (is_call(a, "(")) {
+    m_minus(a[[2]])
```

Comment:
Do you not just want to remove the bracket and do a unary minus here?
```suggestion
    m_uminus(a[[2]])
```

Reply:
yes, thanks - another bug! (the real implementation has actual testing, thankfully)